cas-server-core/src/main/java/org/jasig/cas/authentication/principal/cache/CachingPrincipalAttributesRepository.java

```diff
@@ -0,0 +1,135 @@
+ * based on google guava's caching library.
+ * @author Misagh Moayyed
+ * @since 4.2
+ */
+public final class CachingPrincipalAttributesRepository extends AbstractPrincipalAttributesRepository {
+    private static final long serialVersionUID = 6350244643948535906L;
+    private static final long DEFAULT_MAXIMUM_CACHE_SIZE = 1000;
+
+    private final transient Cache<String, Map<String, Object>> cache;
+    private final PrincipalAttributesCacheLoader cacheLoader =
+            new PrincipalAttributesCacheLoader();
+    /**
+     * Instantiates a new caching attributes principal factory.
+     * Sets the default cache size to {@link #DEFAULT_MAXIMUM_CACHE_SIZE}.
+     * @param attributeRepository the attribute repository
+     * @param timeUnit the time unit
+     * @param expiryDuration the expiry duration
+     */
+    public CachingPrincipalAttributesRepository(final IPersonAttributeDao attributeRepository,
+                                                final TimeUnit timeUnit,
+                                                final long expiryDuration) {
+        this(attributeRepository, DEFAULT_MAXIMUM_CACHE_SIZE, timeUnit, expiryDuration);
+    }
+
+    /**
+     * Instantiates a new caching attributes principal factory.
+     * @param attributeRepository the attribute repository
+     * @param maxCacheSize the max cache size
+     * @param timeUnit the time unit
+     * @param expiryDuration the expiry duration
+     */
+    public CachingPrincipalAttributesRepository(final IPersonAttributeDao attributeRepository,
+                                                final long maxCacheSize,
+                                                final TimeUnit timeUnit,
+                                                final long expiryDuration) {
+        super(attributeRepository, expiryDuration, timeUnit);
+        this.cache = CacheBuilder.newBuilder().maximumSize(maxCacheSize)
+                .expireAfterWrite(expiryDuration, timeUnit).build(this.cacheLoader);
+    }
+
+    @Override
+    protected void addPrincipalAttributesIntoCache(final String id, final Map<String, Object> attributes) {
+        this.cache.put(id, attributes);
+        logger.debug("Cached attributes for {}", id);
+    }
+
+    @Override
+    protected Map<String, Object> getPrincipalAttributesFromCache(final Principal p) {
+        try {
+            return this.cache.get(p.getId(), new Callable<Map<String, Object>>() {
+                @Override
+                public Map<String, Object> call() throws Exception {
+                    logger.debug("No cached attributes could be found for {}", p.getId());
+                    return new HashMap<>();
+                }
+            });
+        } catch (final Exception e) {
+            logger.error(e.getMessage(), e);
+        }
+        return null;
+    }
+
+    @Override
+    public void close() throws IOException {
+        this.cache.cleanUp();
+    }
+
+    @Override
+    public boolean equals(final Object obj) {
```

Comment:
OK. But why redefining both methods just on relying on the parent behaviours? -> `.appendSuper(super.equals(rhs))` and `.appendSuper(super.hashCode())`